### DISCUSSION TITLE
CORS issues with Apollo Server

### DISCUSSION THREAD

**Discussion by theonlydaleking** [vercel/vercel#3998] [General] [answered]
Posted this on spectrum and didn't realise we moved over here. My bad. Reposting: 

i've tried all sorts of shiz for trying to get this to work. Everything gets me the same:
Access to fetch at 'http://myEndpointURI.now.sh/graphql' from origin 'http://localhost:3000' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: Redirect is not allowed for a preflight request.
Current and final now.json that i've got is:
```json
{
  "builds": [
    {
      "src": "./index.ts",
      "use": "@now/node"
    }
  ],
  "routes": [
    {
      "src": "/(.*)",
      "dest": "/index.ts",
      "headers": {
        "Access-Control-Allow-Origin": "*",
        "Access-Control-Allow-Credentials": "true",
        "Access-Control-Allow-Headers": "X-Requested-With, Access-Control-Allow-Origin, X-HTTP-Method-Override, Content-Type, Cache-Control, Authorization, Accept"
      }
    },
    {
      "src": "/.*",
      "methods": ["OPTIONS"],
      "headers": {
        "Access-Control-Allow-Origin": "*",
        "Access-Control-Allow-Credentials": "true",
        "Access-Control-Allow-Headers": "X-Requested-With, Access-Control-Allow-Origin, X-HTTP-Method-Override, Content-Type, Cache-Control, Authorization, Accept"
      },
      "dest": "/cors-handler.ts"
    }
  ]
}
```

the server code is the standard apollo-server stuff. None of the cors options help. It does however work fine on a Digital Ocean VPS:

```ts
import { ApolloServer } from 'apollo-server'
import { schema } from './schema'
import { connectToDatabase } from './Config/config'
require('./Config/config')
const server = new ApolloServer({
  schema,
  playground: true,
  cors: { origin: 'http://localhost:3000' } // this has 0 effect on anything, tried true, false, origins, etc.
})
server.listen().then(async ({ url }) => {
  await connectToDatabase()
  console.log(`🚀  Server running at ${url}`)
})
// module.exports = server
``` 

and my Options Handler: 

```ts
// from here: https://github.com/zeit/now/issues/3015#issuecomment-533435187

export default async function(req: any, res: any) {
  // There's nothing to do here except to end the response since CORS headers
  // are automatically added for all routes using the `now.json` file.
  console.log('options handler being called')
  return res.status(204).send()
}

```
The fact that it works fine on a DO instance leads me to believe its not an apollo problem.

**Comment by tafelito** [comment, marked as answer]
you can't have `"Access-Control-Allow-Origin": "*"` with `"credential": "true"`

check this https://spectrum.chat/zeit/now/cors-policy-block~9871f3e1-652b-4c63-809d-d5aa53ffc176

**Reply by lorstenoplo** [reply]
can u give me a repo link with working code, i'm stuck

**Comment by paulogdm** [comment]
I highly recommend one of the following:

- Set headers in the function itself with `res.setHeader('key', 'value').
- Use the [Headers](https://zeit.co/docs/configuration#project/headers) configuration.
